### ISSUE TITLE
TF 2.11.0/2.12 fails to build in MacOS 13  - XCode 14.1 - issue with ld linker

### ISSUE THREAD

**Issue by feranick** [tensorflow/tensorflow#58368]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Build/Install

### Source

source

### Tensorflow Version

2.11.0-rc

### Custom Code

No

### OS Platform and Distribution

MacOS 13.0

### Python version

3.10.6

### Bazel version

5.3.0

### GCC/Compiler version

clang - XCode 14.1-rc2

### Current Behaviour?

```shell
Compilation fails (log below) under MacOS 13.0 (XCode 14.1-rc2). Compilation works just fine under MacOS 13.x (XCode 14.0.x). Note that compilation fails also for TF 2.10.0 under MacOS 13.0, while it works for MacOS 12.x.
```

### Standalone code to reproduce the issue

```shell
Follow standard compilation from source as indicated here:

https://www.tensorflow.org/install/source
```

### Relevant log output

```shell
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/tensorflow/runtime/archive/4ce3e4da2e21ae4dfcee9366415e55f408c884ec.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/llvm/llvm-project/archive/d8415b02a519f222ecf71b069c96cc85ac635de3.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/protocolbuffers/upb/archive/9effcbcb27f0a665f9f345030188c0b291e32482.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://mirror.bazel.build/github.com/bazelbuild/rules_cc/archive/081771d4a0e9d7d3aa0eed2ef389fa4700dfb23e.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/openxla/stablehlo/archive/fdd47908468488cbbb386bb7fc723dc19321cb83.zip failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/google/XNNPACK/archive/e8f74a9763aa36559980a0c2f37f587794995622.zip failed: class java.io.FileNotFoundException GET returned 404 Not Found
INFO: Analyzed target //tensorflow/tools/pip_package:build_pip_package (544 packages loaded, 39502 targets configured).
INFO: Found 1 target...
ERROR: /Users/feranick/Desktop/tensorflow/tensorflow/python/BUILD:358:27: Linking tensorflow/python/_pywrap_py_exception_registry.so [for host] failed: (Aborted): cc_wrapper.sh failed: error executing command 
  (cd /private/var/tmp/_bazel_feranick/50b852099a3bf3aaa184abce166f8e34/execroot/org_tensorflow && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=13.0 \
    PATH=/opt/usr/local/bin/:/Users/feranick/Documents/Work/c/android-sdk/platform-tools:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/usr/local/MacGPG2/bin:/opt/X11/bin:/Library/Apple/usr/bin \
    XCODE_VERSION_OVERRIDE=14.1.0.14B47b \
    ZERO_AR_DATE=1 \
  external/local_config_cc/cc_wrapper.sh @bazel-out/host/bin/tensorflow/python/_pywrap_py_exception_registry.so-2.params)
# Configuration: 41f578a38c65243e257ab9820de63672c4444f072d1cfdc2d17552cb31aa4b38
# Execution platform: @local_execution_config_platform//:platform
ld: malformed trie, node past end file 'bazel-out/host/bin/_solib_darwin_x86_64/libtensorflow_Spython_S_Upywrap_Utensorflow_Uinternal.so'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error in child process '/usr/bin/xcrun'. 1
external/local_config_cc/cc_wrapper.sh: line 69:  8962 Abort trap: 6           "$(/usr/bin/dirname "$0")"/wrapped_clang "$@"
Target //tensorflow/tools/pip_package:build_pip_package failed to build
ERROR: /Users/feranick/Desktop/tensorflow/tensorflow/python/tools/BUILD:81:10 Middleman _middlemen/tensorflow_Spython_Stools_Sfreeze_Ugraph-runfiles failed: (Aborted): cc_wrapper.sh failed: error executing command 
  (cd /private/var/tmp/_bazel_feranick/50b852099a3bf3aaa184abce166f8e34/execroot/org_tensorflow && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=13.0 \
    PATH=/opt/usr/local/bin/:/Users/feranick/Documents/Work/c/android-sdk/platform-tools:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/usr/local/MacGPG2/bin:/opt/X11/bin:/Library/Apple/usr/bin \
    XCODE_VERSION_OVERRIDE=14.1.0.14B47b \
    ZERO_AR_DATE=1 \
  external/local_config_cc/cc_wrapper.sh @bazel-out/host/bin/tensorflow/python/_pywrap_py_exception_registry.so-2.params)
# Configuration: 41f578a38c65243e257ab9820de63672c4444f072d1cfdc2d17552cb31aa4b38
# Execution platform: @local_execution_config_platform//:platform
INFO: Elapsed time: 13754.646s, Critical Path: 743.77s
INFO: 30482 processes: 7386 internal, 23096 local.
FAILED: Build did NOT complete successfully
```
</details>

**Comment by mohantym** [comment]
@sushreebarsa !
Could you look at this issue.
Thank you!

**Comment by feranick** [comment]
Bug persists with TF 2.11-rc2 and stable release of XCode 14.1.

**Comment by freedomtan** [comment]
FYR, mostly, this is a bug of Xcode 14.x (or more specifically ld? There were some new changes in ld. It failed to build gcc with some early Xcode 14 versions). I ran into this problem since early Xcode 14 betas.

I could build TensorFlow with Xcode 13.x on macOS Monterey (12.x) without problems. When I tried to build it with Xcode 14 on macOS 12.x, it failed.

**Comment by feranick** [comment]
Technically, bazel allows for downloading a "fresh version" of clang upon configuring previous to compilation. But that option is broken. 

https://github.com/tensorflow/tensorflow/issues/46976

**Comment by bhack** [comment]
> Technically, bazel allows for downloading a "fresh version" of clang upon configuring TF previous to compilation. But that option is broken. 
> 
> See related issues: 
> https://github.com/tensorflow/tensorflow/issues/46976
> https://github.com/tensorflow/tensorflow/issues/58453

Yes I don't know what it was broken with this commit https://github.com/tensorflow/tensorflow/issues/58453#issuecomment-1304365920

I think that this configuration it is currently not tested by the CI jobs so that commit was not checked by the CI.

**Comment by feranick** [comment]
Another approach may be to use one of the version of clang from MacPorts. I have several to choose from, but it is not clear how to enable such a version to compile TF in `.bazelrc`. That could at least be a way to test versions that do work. Suggestions are welcome.

**Comment by freedomtan** [comment]
@bhack, @feranick, and @sushreebarsa 

It's not a compiler issue, it's a linker (ld) issue. As I said it seems Apple folks changed ld recently. I managed to build recent master branch with Xcode 14.1 by replacing the ld with the one from Xcode 13.x. My Xcode 14.1 is installed at /Application, I backed up the `ld` in 
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/
```
and replaced it the one from 13.

The ld from Xcode 13.1.6 works for me
```
$ ld -v
@(#)PROGRAM:ld  PROJECT:ld64-764
BUILD 11:29:01 May 17 2022
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 13.1.6, (clang-1316.0.21.2.5) (static support for 28, runtime is 28)
TAPI support using: Apple TAPI version 13.1.6 (tapi-1316.0.7.3)

```
ld from 14.x, e.g, 14.1 doesn't work
```
@(#)PROGRAM:ld  PROJECT:ld64-820.1
BUILD 18:42:42 Sep 11 2022
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 14.0.0, (clang-1400.0.29.202) (static support for 29, runtime is 29)
TAPI support using: Apple TAPI version 14.0.0 (tapi-1400.0.11)
```

Unfortunately, it seems publicly available `ld64` source code is 609,
`https://github.com/apple-oss-distributions/ld64/tree/ld64-609`. I cannot find ld64 7xx and 8xx.

**Comment by flashlabs** [comment]
Hi guys, Xcode [14.2 RC](https://developer.apple.com/news/releases/?id=12072022h) is out. Maybe it will help you?

**Comment by feranick** [comment]
> Hi guys, Xcode [14.2 RC](https://developer.apple.com/news/releases/?id=12072022h) is out. Maybe it will help you?

Thanks. Just tried and it is still broken. Log attached (it's similar to the previous). 
[log.txt](https://github.com/tensorflow/tensorflow/files/10212480/log.txt)

**Comment by freedomtan** [comment]
According to `xcrun  ld -v`, the ld comes with Xcode 14.2 RC is still **ld64-820.1**

**Comment by learning-to-play** [comment]
@kulinseth

**Comment by mihaimaruseac** [comment]
Was this an issue with TF 2.10? (Trying to see if this is a regression or missing support)

**Comment by feranick** [comment]
Yes. The issue is not specific to any version of TF. It started to 
appear in 2.11 simply because of the new XCode that came out at about 
the same time. You can compile either 2.10 or 2.11 with older versions 
of XCode (ld in particular).

**Comment by mihaimaruseac** [comment]
In this case, probably this will need to be handled when the team migrates the MacOS build infrastructure to one that contains the new version of XCode.

Unfortunately, the release process can only run on a single architecture at a time, it will be too expensive to test all combinations. So we could not uncover this before as we're using old `ld` at this point.

PRs to fix this are welcome, if you want to spend some time debugging and fixing this faster than it would take the OSS team to do the migration.

CC-ing @nitins17 too in case there is more available info

**Comment by learning-to-play** [comment]
Please see https://www.tensorflow.org/install/source#macos

**Comment by MichaelHudgins** [comment]
@nitins17 this looks a bit like the other ld build failures we have seen during the build infra migration.  Could you take a quick look.

**Comment by nitins17** [comment]
Not sure if this issue is related but we have seen `ld: malformed trie` errors when trying to build with Xcode command line tools. The issue goes away if built with the full Xcode.

**Comment by feranick** [comment]
> Not sure if this issue is related but we have seen `ld: malformed trie` errors when trying to build with Xcode command line tools. The issue goes away if built with the full Xcode. (see [related comment](https://github.com/tensorflow/tensorflow/issues/57936#issuecomment-1314548038))

Can you please point to a howto guide on using the full XCode for compilation (rather than using the command tools? I'd be happy to try if this works.

**Comment by bzz** [comment]
> Not sure if this issue is related but we have seen `ld: malformed trie` errors when trying to build with Xcode command line tools. The issue goes away if built with the full Xcode. (see [related comment](https://github.com/tensorflow/tensorflow/issues/57936#issuecomment-1314548038))

Thank you for prompt feedback @nitins17! It looks related.

The only difference when building with XCode is that the deployment target/platform version is actually applied to the final binary, as expected from `—macos_minimum_os=11.0` when built with config `macos_arm64`. That can be confirmed with `dyld_info`. Before, it would have been `12.0` and after it is `11.0`.

In my understanding, that disables [linker's new “chained fixups format”](https://mjtsai.com/blog/2021/06/30/faster-app-launching-in-ios-15-and-monterey/) and the same [could be achieved](https://github.com/bazelbuild/bazel/issues/16413#issuecomment-1270603343) with `--linkopt=-Wl,-no_fixup_chains`.

With 12.0 very similar `malformed trie` errors also happen at the runtime when loading (successfully compiled) dynamic libraries from the Python.  (E.g. https://github.com/tensorflow/text/issues/823#issuecomment-1251596272).

What's more cryptic - with 11.0 it would compile&link but not export the symbols properly (`nm` would list them, but `dyld_info -exports` would not) and then just crash the Python process with `SIGBUS`, even with the `lldb` attached, with what looks like some form of OOM on [parsing export entries recursively](https://lldb.llvm.org/cpp_reference/ObjectFileMachO_8cpp.html#ae215c74959752422e3eca3d32bb82b27).

It seems that XCode 14.x just ships a problematic linker, indeed, as downgrading to commandline tools 13.x (`@(#)PROGRAM:ld  PROJECT:ld64-764`) resolves the issue, even if the deployment target/platform version is still 12.0 🤷

**Comment by nitins17** [comment]
> > Not sure if this issue is related but we have seen `ld: malformed trie` errors when trying to build with Xcode command line tools. The issue goes away if built with the full Xcode. (see [related comment](https://github.com/tensorflow/tensorflow/issues/57936#issuecomment-1314548038))
> 
> It seems that XCode 14.x just ships a problematic linker, indeed, as downgrading to commandline tools 13.x (`@(#)PROGRAM:ld PROJECT:ld64-764`) resolves the issue, even if the deployment target/platform version is still 12.0 🤷

Ah, thats unfortunate. Thank you for the insight!

> Can you please point to a howto guide on using the full XCode for compilation (rather than using the command tools? I'd be happy to try if this works.

Set DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer (path to Xcode) and try running the build again.

**Comment by feranick** [comment]
> Set `DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer` (path to Xcode) and try running the build again.

Thank you. Sorry for the silly question, but where should I set that flag? Bash profile? bazelrc?

**Comment by nitins17** [comment]
Either way works, it is an environment variable for specifying the active developer directory for Xcode.  If using a .bazelrc file, include `build --action_env  DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer` in it.

**Comment by freedomtan** [comment]
> > > Not sure if this issue is related but we have seen `ld: malformed trie` errors when trying to build with Xcode command line tools. The issue goes away if built with the full Xcode. (see [related comment](https://github.com/tensorflow/tensorflow/issues/57936#issuecomment-1314548038))
> > 
> > 
> > It seems that XCode 14.x just ships a problematic linker, indeed, as downgrading to commandline tools 13.x (`@(#)PROGRAM:ld PROJECT:ld64-764`) resolves the issue, even if the deployment target/platform version is still 12.0 shrug
> 
> Ah, thats unfortunate. Thank you for the insight!
> 
> > Can you please point to a howto guide on using the full XCode for compilation (rather than using the command tools? I'd be happy to try if this works.
> 
> Set `DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer` (path to Xcode) and try running the build again.

@nitins17 
I don't think that will work. My `xcode-select -p` was `/Applications/Xcode.app/Contents/Developer` (and `xcrun -f ld` reports right ld) and I ran into such problem. If I replace `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld` with ld from Xcode 13.x, then it goes. If setting `DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer`, then something is wrong (e.g., ld other than the one in `/Applications/Xcode.app` is used). I'll try it anyway.

**Comment by freedomtan** [comment]
@nitins17 FYR. Setting `DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer` doesn't work for me on macOS 13.1 + Xcode 14.2 (either M1 or x86_64).

**Comment by feranick** [comment]
The issue still applies to TF 2.12-rc0.

**Comment by feranick** [comment]
I am checking XCode 14.3 beta and it appears that ld is still in version `820.1', so there should be no difference in behavior compared to previous XCode 14.x.

**Comment by yanghired** [comment]
> @bhack, @feranick, and @sushreebarsa
> 
> It's not a compiler issue, it's a linker (ld) issue. As I said it seems Apple folks changed ld recently. I managed to build recent master branch with Xcode 14.1 by replacing the ld with the one from Xcode 13.x. My Xcode 14.1 is installed at /Application, I backed up the `ld` in
> 
> ```
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/
> ```
> 
> and replaced it with the one from 13.
> 
> The ld from Xcode 13.4.1 works for me
> 
> ```
> $ ld -v
> @(#)PROGRAM:ld  PROJECT:ld64-764
> BUILD 11:29:01 May 17 2022
> configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
> LTO support using: LLVM version 13.1.6, (clang-1316.0.21.2.5) (static support for 28, runtime is 28)
> TAPI support using: Apple TAPI version 13.1.6 (tapi-1316.0.7.3)
> ```
> 
> ld from 14.x, e.g, 14.1 doesn't work
> 
> ```
> @(#)PROGRAM:ld  PROJECT:ld64-820.1
> BUILD 18:42:42 Sep 11 2022
> configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
> LTO support using: LLVM version 14.0.0, (clang-1400.0.29.202) (static support for 29, runtime is 29)
> TAPI support using: Apple TAPI version 14.0.0 (tapi-1400.0.11)
> ```
> 
> Unfortunately, it seems publicly available `ld64` source code is 609, `https://github.com/apple-oss-distributions/ld64/tree/ld64-609`. I cannot find ld64 7xx and 8xx. So I don't know what exactly the issue is.

Hi @freedomtan  - thank you for your insights!
Could you please let me know how did you do - "replace the ld file in Xcode 14.x by an old version from Xcode 13.x"? Did you replace other files other than ld, or just the ld file?

Asking because I tried to ONLY replace the ld file, but I got the following info by running ld -v
```
@(#)PROGRAM:ld  PROJECT:ld64-764
BUILD 11:29:07 May 17 2022
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 14.0.0, (clang-1400.0.29.202) (static support for 28, runtime is 29)
TAPI support using: Apple TAPI version 14.0.0 (tapi-1400.0.11)
```
It looks like the ld version is correct, but LLVM and TAPI versions are still incorrect.

Many thanks!

**Comment by feranick** [comment]
As this is now affecting 2 releases (2.11 and the upcoming 2.12), the lease that could be done is to put  a warning on the release notes.

**Comment by freedomtan** [comment]
> Hi @freedomtan - thank you for your insights! Could you please let me know how did you do - "replace the ld file in Xcode 14.x by an old version from Xcode 13.x"? Did you replace other files other than ld, or just the ld file?
> 

replacing the ld is enough. 

> Asking because I tried to ONLY replace the ld file, but I got the following info by running ld -v
> 
> ```
> @(#)PROGRAM:ld  PROJECT:ld64-764
> BUILD 11:29:07 May 17 2022
> configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
> LTO support using: LLVM version 14.0.0, (clang-1400.0.29.202) (static support for 28, runtime is 29)
> TAPI support using: Apple TAPI version 14.0.0 (tapi-1400.0.11)
> ```
> 
> It looks like the ld version is correct, but LLVM and TAPI versions are still incorrect.
> 

That should be fine.

**Comment by feranick** [comment]
Would anyone have a workable version of ld to share? I can't install XCode 13.x once 14.x is in place with MacOS 13.x is installed..

**Comment by kulinseth** [comment]
> Would anyone have a workable version of ld to share? I can't install XCode 13.x once 14.x is in place with MacOS 13.x is installed..

Hi @feranick  this is fixed in https://developer.apple.com/services-account/download?path=/Developer_Tools/Xcode_14.3_beta_2/Xcode_14.3_beta_2.xip.
Can you please check and verify?

**Comment by shkarupa-alex** [comment]
> > Would anyone have a workable version of ld to share? I can't install XCode 13.x once 14.x is in place with MacOS 13.x is installed..
> 
> Hi @feranick this is fixed in https://developer.apple.com/services-account/download?path=/Developer_Tools/Xcode_14.3_beta_2/Xcode_14.3_beta_2.xip. Can you please check and verify?

I've tried with Xcode_14.3_Release_Candidate and now TF 2.12 compilation failed with
```bash
ERROR: /private/var/tmp/_bazel_alex/dc1a9368c8e4ba5b96348c2850b37ab0/external/boringssl/BUILD:161:11: Compiling src/crypto/x509/t_x509.c [for host] failed: (Exit 1): cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics ... (remaining 44 arguments skipped)
external/boringssl/src/crypto/x509/t_x509.c:321:18: error: variable 'l' set but not used [-Werror,-Wunused-but-set-variable]
    int ret = 0, l, i;
                 ^
1 error generated.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
```

**Comment by feranick** [comment]
> > > Would anyone have a workable version of ld to share? I can't install XCode 13.x once 14.x is in place with MacOS 13.x is installed..
> > 
> > 
> > Hi @feranick this is fixed in https://developer.apple.com/services-account/download?path=/Developer_Tools/Xcode_14.3_beta_2/Xcode_14.3_beta_2.xip. Can you please check and verify?
> 
> I've tried with Xcode_14.3_Release_Candidate and now TF 2.12 compilation failed with
> 
> ```shell
> ERROR: /private/var/tmp/_bazel_alex/dc1a9368c8e4ba5b96348c2850b37ab0/external/boringssl/BUILD:161:11: Compiling src/crypto/x509/t_x509.c [for host] failed: (Exit 1): cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics ... (remaining 44 arguments skipped)
> external/boringssl/src/crypto/x509/t_x509.c:321:18: error: variable 'l' set but not used [-Werror,-Wunused-but-set-variable]
>     int ret = 0, l, i;
>                  ^
> 1 error generated.
> Target //tensorflow/tools/pip_package:build_pip_package failed to build
> ```

The same happens with the stable release XCode 14.3, but that is a completely separate issue. With regards of the original issue, `ld` included in this release has finally been updated to version 857 (from 801). I am testing it right now (and I am temporarily overriding the compiler warning above to get to the end of compilation). Will report soon.

**Comment by feranick** [comment]
> 



> I've tried with Xcode_14.3_Release_Candidate and now TF 2.12 compilation failed with
> 
> ```shell
> ERROR: /private/var/tmp/_bazel_alex/dc1a9368c8e4ba5b96348c2850b37ab0/external/boringssl/BUILD:161:11: Compiling src/crypto/x509/t_x509.c [for host] failed: (Exit 1): cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics ... (remaining 44 arguments skipped)
> external/boringssl/src/crypto/x509/t_x509.c:321:18: error: variable 'l' set but not used [-Werror,-Wunused-but-set-variable]
>     int ret = 0, l, i;
>                  ^
> 1 error generated.
> Target //tensorflow/tools/pip_package:build_pip_package failed to build
> ```

I filed a separate issue report for this issue: https://github.com/tensorflow/tensorflow/issues/60191

**Comment by feranick** [comment]
I can confirm that XCode 14.3 with `ld' version 857 now works, and compilation is successful. I am going to close this as resolved.

**Comment by feranick** [comment]
For those interested, to solve the unrelated issue above, you can temporarily fix it by removing the unused variable, external/boringssl/src/crypto/x509/t_x509.c (see diff).
[issue_60191.txt](https://github.com/tensorflow/tensorflow/files/11128466/issue_60191.txt)
